api/pipeline_spec.proto

```diff
@@ -0,0 +1,332 @@
+      INT = 1;
+      DOUBLE = 2;
+      STRING = 3;
+    }
+    // Required field. The type of the runtime parameter.
+    Type type = 1;
+    // Optional field. Default value of the runtime parameter. If not set and
+    // the runtime parameter value is not provided during runtime, an error will
+    // be raised.
+    Value default_value = 2;
+  }
+  // The runtime parameters of the pipeline. Keyed by parameter name.
+  map<string, RuntimeParameter> runtime_parameters = 6;
+}
+
+// The spec of a pipeline task.
+message PipelineTaskSpec {
+  // Basic info of a pipeline task.
+  PipelineTaskInfo task_info = 1;
+
+  // The specification of a task input.
+  message InputSpec {
+    // The name of the producer task.
+    string producer_task = 1;
+
+    // The key of the output.
+    string output_key = 2;
+  }
+
+  // Specification for task inputs which connect to the outputs from
+  // upstream tasks. Keyed by input name.
+  map<string, InputSpec> inputs = 2;
+
+  // The specification of a task output.
+  message OutputSpec {
+    // The type of the artifact.
+    PipelineTypeSchema artifact_type = 1;
+
+    // The properties of the artifact.
+    map<string, PipelineValueOrRuntimeParameter> properties = 2;
+
+    // The custom properties of the artifact.
+    map<string, PipelineValueOrRuntimeParameter> custom_properties = 3;
+  }
+
+  // Specification for task outputs. Keyed by output name.
+  map<string, OutputSpec> outputs = 3;
+
+  // Specification for task parameters, which contains the values provided by
+  // the pipeline users or from runtime parameters. Keyed by parameter name.
+  map<string, PipelineValueOrRuntimeParameter> parameters = 4;
+
+  // Label for the executor of the task.
+  // The specification will be specified in the deployment config.
+  // For example:
+  // ```
+  // tasks:
+  // - task_info:
+  //     name: trainer
+  //   executor_label: trainer
+  // deployment_config:
+  //   @type: cloud.ml.pipelines.v1alpha3.proto.PipelineDeploymentConfig
+  //   executors:
+  //     trainer:
+  //       container:
+  //         image: gcr.io/tfx:latest
+  //         args: []
+  // ```
+  string executor_label = 5;
+
+  // Names of the upstream tasks of the current task. The current task will
+  // start until all upstream tasks are finished. [PipelineTaskSpec.inputs]()
+  // defines the data dependency of the task, which will also imply a task
+  // dependency on the producer task. There is no need to define two depdencies
+  // to the same upstream task in this case.
+  repeated string upstream_tasks = 6;
+
+  message CachingOptions {
+    // Whether or not to enable cache for this task.
+    bool enable_cache = 1;
+  }
+  CachingOptions caching_options = 7;
+}
+
+// Basic info of a pipeline.
+message PipelineInfo {
+  // Required field. The name of the pipeline.
+  // The name will be used to create or find pipeline context in MLMD.
+  string name = 1;
+}
+
+// The definition of a type in MLMD.
+message PipelineTypeSchema {

```

Comment:
This is an Artifact Type Schema to store in MLMD - better to use the same name?

Reply:
Renamed it to ArtifactTypeSchema.